src/anomalib/metrics/per_image/binclf_curve_numpy.py

```diff
@@ -0,0 +1,435 @@
+
+    if scores_batch.dtype.kind != "f":
+        msg = (
+            "Expected `scores_batch` to be an floating ndarray with anomaly scores_batch,"
+            f" but got ndarray with dtype {scores_batch.dtype}"
+        )
+        raise TypeError(msg)
+
+    if scores_batch.ndim != 2:
+        msg = f"Expected `scores_batch` to be 2D, but got {scores_batch.ndim}"
+        raise ValueError(msg)
+
+
+def _validate_gts_batch(gts_batch: ndarray) -> None:
+    """gts_batch (ndarray): boolean (N, D)."""
+    if not isinstance(gts_batch, ndarray):
+        msg = f"Expected `gts_batch` to be an ndarray, but got {type(gts_batch)}"
+        raise TypeError(msg)
+
+    if gts_batch.dtype.kind != "b":
+        msg = (
+            "Expected `gts_batch` to be an boolean ndarray with anomaly scores_batch,"
+            f" but got ndarray with dtype {gts_batch.dtype}"
+        )
+        raise TypeError(msg)
+
+    if gts_batch.ndim != 2:
+        msg = f"Expected `gts_batch` to be 2D, but got {gts_batch.ndim}"
+        raise ValueError(msg)
+
+
+# =========================================== PYTHON VERSION ===========================================
+
+
+def _binclf_one_curve_python(scores: ndarray, gts: ndarray, threshs: ndarray) -> ndarray:
+    """ONE binary classification matrix at each threshold (PYTHON implementation).
+
+    In the case where the thresholds are given (i.e. not considering all possible thresholds based on the scores),
+    this weird-looking function is faster than the two options in `torchmetrics` on the CPU:
+        - `_binary_precision_recall_curve_update_vectorized`
+        - `_binary_precision_recall_curve_update_loop`
+
+    (both in module `torchmetrics.functional.classification.precision_recall_curve` in `torchmetrics==1.1.0`).
+
+    ATTENTION: VALIDATION IS NOT DONE HERE. Make sure to validate the arguments before calling this function.
+
+    Args:
+        scores (ndarray): Anomaly scores (D,).
+        gts (ndarray): Binary (bool) ground truth of shape (D,).
+        threshs (ndarray): Sequence of thresholds in ascending order (K,).
+
+    Returns:
+        ndarray: Binary classification matrix curve (K, 2, 2)
+
+        See docstring of `binclf_multiple_curves` for details.
+    """
+    num_th = len(threshs)
+
+    # POSITIVES
+    scores_positives = scores[gts]
+    # the sorting is very important for the algorithm to work and the speedup
+    scores_positives = np.sort(scores_positives)
+    # variable updated in the loop; start counting with lowest thresh ==> everything is predicted as positive
+    num_pos = current_count_tp = scores_positives.size
+    tps = np.empty((num_th,), dtype=np.int64)
+
+    # NEGATIVES
+    # same thing but for the negative samples
+    scores_negatives = scores[~gts]
+    scores_negatives = np.sort(scores_negatives)
+    num_neg = current_count_fp = scores_negatives.size
+    fps = np.empty((num_th,), dtype=np.int64)
+
+    def score_less_than_thresh(thresh):  # noqa: ANN001, ANN202
```

Comment:
I get why we need this but is there a way to avoid inline methods? Maybe a combination of lambda and partial functions.

Reply:
i believe ruff doesnt like lambdas, but i changed to use a partial